src/lonlat.rs

```diff
@@ -42,50 +48,69 @@ impl Latitude {
-        let (min, only_spaces) = parse_bytes_trailing_spaces(&[b[2], b[3]], only_spaces)
+        total_spaces += num_spaces;
+        let (min_frac, num_spaces) = parse_bytes_trailing_spaces(&[b[5], b[6]], num_spaces > 0)
             .ok_or_else(|| AprsError::InvalidLatitude(b.to_owned()))?;
+        total_spaces += num_spaces;
 
-        let (min_frac, _) = parse_bytes_trailing_spaces(&[b[5], b[6]], only_spaces)
+        let precision = Precision::from_num_digits(total_spaces)
             .ok_or_else(|| AprsError::InvalidLatitude(b.to_owned()))?;
 
         let value = deg as f64 + min as f64 / 60. + min_frac as f64 / 6_000.;
         let value = if north { value } else { -value };
 
-        Self::new(value).ok_or_else(|| AprsError::InvalidLatitude(b.to_owned()))
+        let lat = Self::new(value).ok_or_else(|| AprsError::InvalidLatitude(b.to_owned()))?;
+
+        Ok((lat, precision))
     }
 
     pub(crate) fn parse_compressed(b: &[u8]) -> Result<Self, AprsError> {
         let value = 90.0
             - (base91::decode_ascii(b).ok_or_else(|| AprsError::InvalidLatitude(b.to_owned()))?
                 / 380926.0);
-        if value > 90. || value < -90. {
-            return Err(AprsError::InvalidLatitude(b.to_owned()));
-        }
 
-        Ok(Self(value))
+        Self::new(value).ok_or_else(|| AprsError::InvalidLatitude(b.to_owned()))
     }
 
     pub(crate) fn encode_compressed<W: Write>(&self, buf: &mut W) -> Result<(), EncodeError> {
         let value = (90.0 - self.0) * 380926.0;
         base91::encode_ascii(value, buf, 4)
     }
 
-    pub(crate) fn encode_uncompressed<W: Write>(&self, buf: &mut W) -> Result<(), EncodeError> {
+    pub(crate) fn encode_uncompressed<W: Write>(
+        &self,
+        buf: &mut W,
+        precision: Precision,
+    ) -> Result<(), EncodeError> {
         let lat = self.0;
 
-        if !(-90.0..=90.0).contains(&lat) {
-            return Err(EncodeError::InvalidLatitude(lat));
-        }
-
         let (dir, lat) = if lat >= 0.0 { ('N', lat) } else { ('S', -lat) };
 
         let deg = lat as u32;
         let min = ((lat - (deg as f64)) * 60.0) as u32;
         let min_frac = ((lat - (deg as f64) - (min as f64 / 60.0)) * 6000.0).round() as u32;
 
-        write!(buf, "{:02}{:02}.{:02}{}", deg, min, min_frac, dir)?;
+        // into_bytes is safe because we know the string only contains ASCII values
+        let mut digit_buffer = format!("{:02}{:02}{:02}", deg, min, min_frac).into_bytes();
+
+        // zero out fields as required for precision
+        // Ideally we would be doing some clever rounding here
+        // E.g. if last 2 digits were blanked,
+        // 4905.83 would become 4906.__
+        for i in (6 - precision.num_digits())..6 {
+            digit_buffer[i as usize] = b' ';
+        }
+
+        buf.write_all(&digit_buffer[0..4])?;
+        write!(buf, ".")?;
+        buf.write_all(&digit_buffer[4..6])?;
+        write!(buf, "{}", dir)?;
         Ok(())
```

Comment:
I truly hate this. It should work fine but it looks bad, and will need to be rewritten anyway for no-std (I plan on doing this eventually so that this library can be used in embedded systems)

Reply:
Yeah... I don't have a better idea yet, but will think about it.